src/sage/rings/species.py

```diff
@@ -0,0 +1,475 @@
+    def __init__(self, parent, C, ambient_group):
+        r"""
+        A conjugacy class of directly indecomposable subgroups.
+        """
+        Element.__init__(self, parent)
+        self._C = C
+        self._smallgen = C.gens_small()
+        self._ambient_group = ambient_group
+
+    @cached_method
+    def subgroup_of(self):
+        r"""
+        Return the group which this conjugacy class of
+        directly indecomposable subgroups belongs to.
+        """
+        return self._ambient_group
+
+    @cached_method
+    def _element_key(self):
+        r"""
+        Return the key for this element.
+        """
+        return tuple([self._C.degree(), self._C.order(),
+                      self._ambient_group])
+
+    def __hash__(self):
+        return hash(self._element_key)
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+        """
+        return f"{self._smallgen}"
+
+    def __eq__(self, other):
+        r"""
+        Return if ``self`` is equal to ``other``.
+
+        Two directly indecomposable subgroups belong to the same conjugacy
+        class if they have the same ambient group and are conjugate within it.
+        """
+        # No __le__ method because does it really make sense to have one here?
+        return (isinstance(other, ConjugacyClassOfDirectlyIndecomposableSubgroups)
+                and self._ambient_group == other._ambient_group
+                and _is_conjugate(self._ambient_group, self._C, other._C))
+
+class ConjugacyClassesOfDirectlyIndecomposableSubgroups(UniqueRepresentation, Parent, ElementCache):
+
+    def __init__(self):
+        r"""
+        Conjugacy classes of directly indecomposable subgroups.
+        """
+        Parent.__init__(self, category=InfiniteEnumeratedSets())
+        ElementCache.__init__(self)
+    
+    def _element_constructor_(self, x):
+        r"""
+        ``x`` is an element of ``self`` or a tuple `(H, G)` such that
+        `H` is directly indecomposable and `G` is the ambient group of
+        `H`.
+        """
+        if parent(x) == self:
+            return x
+        H, G = x
+        # can't handle any normalization before this point because
+        # of subgroup check
+        if H.is_subgroup(G):
+            if len(H.disjoint_direct_product_decomposition()) > 1:
+                raise ValueError(f"{H} is not directly indecomposable")
+            # do domain normalization here?
+            elm = self.element_class(self, H, G)
+            return self._cache_get(elm)
+        raise ValueError(f"{H} is not a subgroup of {G}")
+    
+    Element = ConjugacyClassOfDirectlyIndecomposableSubgroups
+
+class AtomicSpeciesElement(Element):
+    def __init__(self, parent, dis, mc):
+        r"""
+        An atomic species. ``dis`` is an instance of
+        ConjugacyClassOfDirectlyIndecomposableSubgroups
+        and ``mc`` is the multicardinality of the atomic
+        species.
```

Comment:
The multicardinality is not enough, you also need the tuple of set partitions, that tell us which subsets of the domain belong to which variable.

Reply:
But I'm normalizing the generators such that `[1,..,a]` is in variable 1, `[a+1,...b]` is in variable 2, so on. That should remove the need for storing the set partitions explicitly, right? Since we know how to recover it from the multicardinality. Also, I'm doing this way so I can construct $S_{n_1} \times S_{n_2} \times \ldots \times S_{n_k}$ simply using `SymmetricGroup(n).young_subgroup([n_1, n_2, ... , n_k])` to check if the provided group is a valid subgroup.

Reply:
oh no

I see.

Reply:
If the domain partition information is not passed to `ConjugacyClassesOfD....`, then when we try to find an element in the cache, we are checking if another group that is its conjugate within $S_n$ already exists. If that is the case, then we are going to end up with the wrong generators anyway, if a conjugate group is returned.

Ideas to fix this:
1. Pass the domain partition information to  `ConjugacyClassesOfD....`. Not favourable, that class shouldn't have to know about it.
2. When we get the returned element, we use `libgap.RepresentativeAction` on our original (normalized) group and the returned group with $S_n$ as the parent group to find the conjugating element and apply it to our domain partition.

Reply:
I am in favour of the second option.

I wonder whether, given a directly indecomposable group $G$, there is a unique (!) finest integer partition $\lambda_1 \geq \lambda_2\geq \dots \lambda_\ell > 0$ such that $G$ is a subgroup of $\mathfrak S_{\lambda_1} \times \mathfrak S_{\lambda_2}\times \dots\times \mathfrak S_{\lambda_\ell}$.  That would be a nice further step into normalization.

Reply:
I think the answer to my question is "yes" and the partition is not very difficult to find.  Let $C$ be the set of sets whose element appear in a cycle in any of the generators of $G$.  For example, if the generators are 

$(1,3)(4,7)$, $(2,5)(6,8)$ and $(1,4)(2,5)(3,7)$,

then $C$ is the set $\\{\\{1,3\\}, \\{4,7\\}, \\{2,5\\}, \\{6,8\\}, \\{1,4\\}, \\{3,7\\}\\}$.

Then the smallest Young subgroup of the symmetric group which has $G$ as a subgroup is obtained by taking the union of those sets which have non-empty intersection.  In the example above, this would be the group $\mathcal S_{\\{1,3,4,7\\}}\times\mathcal S_{\\{2,5\\}}\times\mathcal S_{\\{6,8\\}}$.

Thus, we would rewrite the set of generators of $G$ by mapping $\\{1,3,4,7\\}$ to $\\{1,2,3,4\\}$, $\\{2,5\\}$ to $\\{5,6\\}$ and $\\{6,8\\}$ to $\\{7,8\\}$, and obtain

$(1,2)(3,4)$, $(1,3)(2,4)(5,6)$, and $(5,6)(7,8)$

as "beautified" set of generators.

I guess it's not super-important to do this, but it may make the groups a little easier to understand and is very cheap.

Reply:
Didn't you just describe the orbits of `G`?

Reply:
Yes 😂

Reply:
This is now incorporated.